lib/dry/operation/extensions/rom.rb

```diff
@@ -0,0 +1,118 @@
+end
+
+module Dry
+  class Operation
+    module Extensions
+      # Add rom transaction support to operations
+      #
+      # When this extension is included, you can use a `#transaction` method
+      # to wrap the desired steps in a rom transaction. If any of the steps
+      # returns a `Dry::Monads::Result::Failure`, the transaction will be rolled
+      # back and, as usual, the rest of the flow will be skipped.
+      #
+      # The extension expects the including class to give access to the rom
+      # container via a `#rom` method.
+      #
+      # ```ruby
+      # class MyOperation < Dry::Operation
+      #   include Dry::Operation::Extensions::ROM
+      #
+      #   attr_reader :rom
+      #
+      #   def initialize(rom:)
+      #     @rom = rom
+      #   end
+      #
+      #   def call(input)
+      #     attrs = step validate(input)
+      #     user = transaction do
+      #       new_user = step persist(attrs)
+      #       step assign_initial_role(new_user)
+      #       new_user
+      #     end
+      #     step notify(user)
+      #     user
+      #   end
+      #
+      #   # ...
+      # end
+      # ```
+      #
+      # By default, the `:default` gateway will be used. You can change this
+      # when including the extension:
+      #
+      # ```ruby
+      # include Dry::Operation::Extensions::ROM[gateway: :my_gateway]
+      # ```
+      #
+      # Or you can change it at runtime:
+      #
+      # ```ruby
+      # user = transaction(gateway: :my_gateway) do
+      #  # ...
+      # end
+      # ```
+      #
+      # @see https://rom-rb.org
+      module ROM
+        DEFAULT_GATEWAY = :default
+
+        # @!method transaction(gateway: DEFAULT_GATEWAY, &steps)
+        #  Wrap the given steps in a rom transaction.
+        #
+        #  If any of the steps returns a `Dry::Monads::Result::Failure`, the
+        #  transaction will be rolled back and `:halt` will be thrown with the
+        #  failure as its value.
+        #
+        #  @yieldreturn [Object] the result of the block
+        #  @raise [Dry::Operation::InterfaceNotImplementedError] if the including
+        #    class doesn't define a `#rom` method.
+        #  @see Dry::Operation#steps
+
+        def self.included(klass)
+          klass.include(self[])
+        end
+
+        # Include the extension providing a custom gateway
+        #
+        # @param gateway [Symbol] the rom gateway to use
+        def self.[](gateway: DEFAULT_GATEWAY)
+          Builder.new(gateway: gateway)
+        end
+
+        # @api private
+        class Builder < Module
+          def initialize(gateway:)
+            super()
+            @gateway = gateway
+          end
+
+          def included(klass)
+            class_exec(@gateway) do |default_gateway|
+              klass.define_method(:transaction) do |gateway: default_gateway, &steps|
+                raise Dry::Operation::InterfaceNotImplementedError, <<~MSG unless respond_to?(:rom)
```

Comment:
Hmmm I see your point. On the other side, `ExtensionError` sounds to vague to me. If we want to separate between core vs extension errors, maybe we could nest within the extensions namespace? In this case,`Dry::Operation::Extensions::InterfaceNotImplementedError`. WDYT?

Reply:
@waiting-for-dev Good question, and I have several thoughts :)

IMO, `ExtensionError` being _somewhat_ vague is a feature, not a bug :) It's a short name that makes it immediately clear the error is something to do with an extension. And when the exception is raised, it will carry a message that makes the specific details of the error clear.

Having a more generic error class like this means we can use it in the future if we need to raise other issues to do with extensions. I've taken a similar approach with e.g. `Hanami::SliceLoadError` and I've found it quite freeing. Having this has meant I can raise this error in appropriate circumstances without wringing my hands over trying to come up with hyper-specific error names (or burdening the user with trying to read or reason about them). Remember: we always have the exception's message to explain ourselves more fully.

When do I think we should turn to more specific or single-circumstance error classes? IMO probably a couple of reasons: (a) when an error or a category of errors has become so distinct that it needs its own name, or (b) when those errors need to carry additional attributes that are important for programatically interacting with them.

In the case of a new gem, I think I'd rather start with simpler, looser names and then iterate towards more specific ones when we learn from user feedback, or our own experience in continuing to build out the gem.

Are you OK if we stuck with `ExtensionError` here?

Reply:
Looks good, @timriley. Changed and merging 🙂